### ISSUE TITLE
Unable to deploy on GKE, blocked at bookie.yaml step

### ISSUE THREAD

**Issue by pixime** [apache/pulsar#6135]
I'm trying to deploy pulsar on gke according to the Deploy pulsar components chapter in the documentation (https://pulsar.apache.org/docs/en/deploy-kubernetes/#deploy-pulsar-components). I use the apache-pulsar-2.5.0/deployment/kubernetes/google-kubernetes-engine yaml source directory. 

Steps :
- kubectl create -f zookeeper.yaml : OK, ok 3 replicated pods are running
- kubectl create -f cluster-matadata.yaml : OK, completed
- Kubectl create -f bookie.yaml failed.

I got the following error many times in the bookie-autorecovery logs: Opening socket connection to server localhost/127.0.0.1:2181. Will not attempt to authenticate using SASL (unknown error). 

As the zookeeper client StatefulSetport  is set to 2181, I suppose that the bookie-autorecovery is unable to connect to zookeeper. Naturally it ends by a CrashLoopBackOff.

Notice also that the bookie pods seem to be stuck, no logs.

My gke configuration is:
- 3 nodes
- n1-standard-4
- no autoscaling at the moment
- europe-west1

Any idea ?

**Comment by gaborho** [comment]
Hello, I have the same issue, but on AWS EKS. In case you remove PULSAR_PREFIX_ from all the keys defined in the ConfigMap of resource of bookie.yaml (in case of GKE) or in bookkeeper.yaml (in case of AWS) then it works.

**Comment by pixime** [comment]
What do you mean exactly ?

This is what I did:

**1. fist attempt:**
I commented only the PULSAR_PREFIX_zkServers variable
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: bookie-config
data:
  PULSAR_MEM: "\" -Xms4g -Xmx4g -XX:MaxDirectMemorySize=4g\""
  PULSAR_GC: "\" -XX:+UseG1GC \""
  PULSAR_PREFIX_dbStorage_writeCacheMaxSizeMb: "1024"
  PULSAR_PREFIX_dbStorage_readAheadCacheMaxSizeMb: "1024"
  #PULSAR_PREFIX_zkServers: zk-0.zookeeper,zk-1.zookeeper,zk-2.zookeeper
  PULSAR_PREFIX_statsProviderClass: org.apache.bookkeeper.stats.prometheus.PrometheusMetricsProvider
  PULSAR_PREFIX_useHostNameAsBookieID: "true"
```
**But it failed**.

**2° attempt:**
With all PULSAR_PREFIX variables commented
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: bookie-config
data:
  PULSAR_MEM: "\" -Xms4g -Xmx4g -XX:MaxDirectMemorySize=4g\""
  PULSAR_GC: "\" -XX:+UseG1GC \""
  #PULSAR_PREFIX_dbStorage_writeCacheMaxSizeMb: "1024"
  #PULSAR_PREFIX_dbStorage_readAheadCacheMaxSizeMb: "1024"
  #PULSAR_PREFIX_zkServers: zk-0.zookeeper,zk-1.zookeeper,zk-2.zookeeper
  #PULSAR_PREFIX_statsProviderClass: org.apache.bookkeeper.stats.prometheus.PrometheusMetricsProvider
  #PULSAR_PREFIX_useHostNameAsBookieID: "true"
```
**And it failed as well**.

**Comment by gaborho** [comment]
What I meant is that you should remove PULSAR_PREFIX_ from the key names. So you should try with this settings:

dbStorage_writeCacheMaxSizeMb: "1024"
dbStorage_readAheadCacheMaxSizeMb: "1024"
zkServers: zk-0.zookeeper,zk-1.zookeeper,zk-2.zookeeper
statsProviderClass: org.apache.bookkeeper.stats.prometheus.PrometheusMetricsProvider
useHostNameAsBookieID: "true"

**Comment by hjohannsen** [comment]
We're facing the same issue with AWS EKS. 

The [apply-config-from-env.py](https://github.com/apache/pulsar/blob/master/docker/pulsar/scripts/apply-config-from-env.py) script doesn't strip away the `PULSAR_PREFIX_` when overwriting _existing_ keys. 
So, if a key, such as `zkServers` already exists in the `bookkeeper.conf`, the setting isn't overwritten. 

In the [EKS deployment sources](https://github.com/apache/pulsar/tree/master/deployment/kubernetes/aws) which we departed from, the `zkServers` env variable is defined _with_ the prefix (`PULSAR_PREFIX_zkServers`). And in the config from the Docker image the setting already exists: `zkServers=localhost:2181`. As a consequence, overwriting doesn't happen and the Bookkeeper looks for the Zookeeper at `localhost` instead of `zk-x.zookeeper`.

Our temporary workaround is to duplicate the settings in the `ConfigMap`, with and without prefix.

Is there any reasoning behind ignoring the prefix for existing keys or is it just a bug?

**Comment by jiazhai** [comment]
@hjohannsen @gaborho @pixime , This is a bug and should be fixed in PR #5790 
Would you please help verify in the latest release 2.5.0?

**Comment by jiazhai** [comment]
Oh, seems @pixime is using 2.5.0. @wolfstudy Would you please help check if there is any missing fix.

**Comment by hjohannsen** [comment]
@jiazhai, thanks for the hint! Sure I can help verify. 

Just to make sure I fully understand how it's _supposed_ to work:

**Background**
I was trying to set up Pulsar with EKS, simply following the instructions in ["Deploying Pulsar on Kubernetes"](https://pulsar.apache.org/docs/v1.22.0-incubating/deployment/Kubernetes/).
For that purpose I cloned the Pulsar repo and copied the source files from `deployment/kubernetes/aws`. At that time, the `.yaml` files had the `PULSAR_PREFIX_` introducing most of the properties in the `ConfigMap`.
From the [apply-config-from-env.py](https://github.com/apache/pulsar/blob/master/docker/pulsar/scripts/apply-config-from-env.py) script, I inferred that the script wouldn't apply the settings to properties that were _already set_ in the relevant config file (in the Docker container). Because of the prefix.
Given that constellation, you would have to
- omit the prefix for properties already defined in the config,
- add the prefix for new properties absent from the config. 

**Recent Changes**
Looking at the history of, say `bookkeeper.yaml`, I see that the prefix was added consistently in `d74c1acb` (#5790) and removed again in `d631156d` (Issue #6192).  

Nothing changed in  [apply-config-from-env.py](https://github.com/apache/pulsar/blob/master/docker/pulsar/scripts/apply-config-from-env.py). So I still need to add the prefix to set properties that aren't preconfigured in the config file.

**Question**
Am I right in assuming that I shouldn't need to worry, at all, about settings that aren't preconfigured in the config from the Docker image? I only ever need to overwrite existing properties without ever adding a prefix?
(I'd be totally fine with that.)
